### ISSUE TITLE
[JIT] Fusion of Dropout without constant is_training parameter is unsuccessful

### ISSUE THREAD

**Issue by kevinstephano** [pytorch/pytorch#24032]
## 🐛 Bug

When I create a jit.script function that includes `torch.nn.functional.dropout` without a constant `is_training` parameter.  The fusion does not work.  This did previously work.

## To Reproduce
Here is an example script:

```
import torch
import torch.nn.functional as F

@torch.jit.script
def jit_dropout_add(x, residual, prob, is_training) :
    # type: (Tensor, Tensor, float, bool) -> Tensor
    out = F.dropout(x, p=prob, training=is_training)
    out = residual + out
    return out

@torch.jit.script
def jit_dropout_add_const(x, residual, prob) :
    # type: (Tensor, Tensor, float) -> Tensor
    out = F.dropout(x, p=prob, training=True)
    out = residual + out
    return out


inputs = torch.ones(5, 5 , dtype=torch.float16, device=torch.device("cuda:0"), requires_grad=True)
residuals = torch.ones(5, 5 , dtype=torch.float16, device=torch.device("cuda:0"))

output_bad = jit_dropout_add(inputs, residuals, 0.1, True)
output_good = jit_dropout_add_const(inputs, residuals, 0.1)
```

## Expected behavior

If you run the script above with `PYTORCH_FUSION_DEBUG=1`, you will see in the good case, shown below, `n0` and `n1` read from memory.

```
void kernel_1(IndexType totalElements, const TensorInfo<half,1> t0, const TensorInfo<half,1> t1, double s2, double s3, const TensorInfo<half,1> t4, const TensorInfo<half,1> t5, const TensorInfo<uint8_t,1> t6 ,unsigned long long seed, unsigned long long offset) {
  
  for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
        linearIndex < totalElements;
        linearIndex += gridDim.x * blockDim.x) {
      

      // calculate the results
      float n0 = __half2float(t0.data[t0_offset]);
      float n1 = __half2float(t1.data[t1_offset]);
      double n2 = s2;
      double n3 = s3;
      int64_t n4 = 1;
      float n5 = uniform(rnd());
      uint8_t n6 = n5 < n3;
      float n7 = (((float) n6));
      float n8 = n7 * n1;
      float n9 = n8 * ((float) n2);
      float n10 = n0 + ((float) n4)*n9;
      t4.data[t4_offset] = __float2half(n10);
      t5.data[t5_offset] = __float2half(n9);
      t6.data[t6_offset] = n6;
    
    }
}
```

cc @suo

**Comment by eellison** [comment]
I think this is the result of https://github.com/pytorch/pytorch/pull/20760 @suo.

**Comment by ptrblck** [comment]
@suo Do you know, if this is the result of #20760?

**Comment by suo** [comment]
It certainly seems so. @ailzhang did you have a recent change that fixed this or is it an unrelated dropout fusion thing?

**Comment by ailzhang** [comment]
I think this is likely fixed on master. I'm building locally to confirm.

**Comment by kevinstephano** [comment]
I am still seeing this as an issue. It seems like this might have not been resolved?

**Comment by wconstab** [comment]
@wconstab move to PE board

**Comment by Krovatkin** [comment]
thanks @wconstab !!!

**Comment by eellison** [comment]
I would suggest removing the `if train` logic in symbolic_shapes and then when we guard DifferentiableGraphs add a check that the training attribute of dropout is Truee

**Comment by kevinstephano** [comment]
I should note we have a solution for this problem that we haven't tried to upstream, yet.  We put the training parameter into the dropout parameters and made a cleaner implementation in ATen like native_layer_norm as an example.

**Comment by desertfire** [comment]
@kevinstephano Is the plan to upstream your fix?

**Comment by csarofeen** [comment]
I believe this is the associated PR: https://github.com/pytorch/pytorch/pull/63937 @jjsjann123 and @kevinstephano to confirm.

**Comment by janvainer** [comment]
HI this issue is still failing with the latest torch release. Any update?

**Comment by jjsjann123** [comment]
Err. I think the issue might have come from what NNC supports.

Running on latest master, I do see `native_dropout` in a plain differentiable graph that can be fused with the binary add. (switching nvfuser it actually produces a fused graph). So the original problem is tackled.

cc'ing @eellison on NNC coverage.

**Comment by eellison** [comment]
According to @ZolotukhinM NNC doesn't support random so it's not just op coverage but more fundamental. The alternative for now is 
`with torch.jit.fuser("fuser2")` until a future release when nvfuser is on by default.

**Comment by jjsjann123** [comment]
In that sense I'm closing this bug since nnc coverage seems to be a separate issue and discussion above could be distracting.
